### ISSUE TITLE
AL server crashing with new AL Compiler

### ISSUE THREAD

**Issue by SirBETE** [microsoft/AL#5773]
**Describe the bug**
Opening VSC with the new AL Compiler loaded in, won't activate it. It keeps crashing.

[Info  - 14:47:12] Connection to server got closed. Server will restart.
Unhandled exception. System.TypeInitializationException: The type initializer for 'Microsoft.Dynamics.Nav.Deployment.Telemetry.TelemetryServiceManager' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.
 ---> System.ArgumentException: External component has thrown an exception.
   at System.Globalization.CompareInfo.CompareString(ReadOnlySpan`1 string1, ReadOnlySpan`1 string2, CompareOptions options)
   at System.Globalization.CompareInfo.Compare(String string1, String string2, CompareOptions options)
   at System.Globalization.TextInfo.PopulateIsAsciiCasingSameAsInvariant()
   at System.Globalization.TextInfo.ChangeCaseCommon[TConversion](String source)
   at System.Globalization.TextInfo.ToUpper(String str)
   at System.String.ToUpperInvariant()
   at System.Diagnostics.Tracing.EventSource.GetGuid(Type eventSourceType)
   at System.Diagnostics.Tracing.EventSource..ctor(EventSourceSettings settings, String[] traits)
   at System.Diagnostics.Tracing.EventSource..ctor()
   at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..ctor()
   at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..cctor()
   --- End of inner exception stack trace ---
   at Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.DequeueAndSend(TimeSpan timeout)
   at Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.Runner()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   --- End of inner exception stack trace ---
   at Prod.EditorServices.Host.Program.Main(String[] args) in D:\a\1\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Host\Program.cs:line 135
[Error - 14:47:12] Connection to server got closed. Server will not be restarted.

On my server I also have VS Installed (for report design).
Microsoft Visual Studio Community 2019
Version 16.5.2
VisualStudio.16.Release/16.5.2+29926.136
Microsoft .NET Framework
Version 4.7.03062

This crash has something to do with Microsoft.AspNetCore.Identity nuget package, where's a bug in v 3.1.1 (and fixed in 3.1.2). So I first upgraded VS to this listed version, then installed the latest version of  https://www.nuget.org/packages/Microsoft.AspNetCore.Identity.EntityFrameworkCore/3.1.3  to a test C# project.

But the AL Compiler keeps crashing ...

Any ideas how to get the new AL Compiler working?

**To Reproduce**
Steps and to reproduce the behavior:
1. open VSC with AL compiler active

``` AL Code to reproduce the issue ```
/
**Expected behavior**
Compiler us active

**Screenshots**

**5. Versions:**

 - AL Language: 5.0.254558

 - Business Central: 15.2

https://www.yammer.com/dynamicsnavdev/threads/618431748030464

**Comment by SirBETE** [comment]
cmd dotnet --info gives:
C:\Windows\system32>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  3.1.201 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

Some more googling gives:
https://github.com/dotnet/Scaffolding/issues/454

Extrapolating this to dotnet 3.1, means I need packageversion 3.1 too
Install-Package Microsoft.VisualStudio.Web.CodeGeneration.Design -Version 3.1.2
Installed it into a test project, still same error

**Comment by SirBETE** [comment]
https://stackoverflow.com/questions/43330915/could-not-load-file-or-assembly-microsoft-build-frameworkvs-2017
suggest 
> Our local build script was using an older version of nuget.exe (4.7.1.5393) to restore NuGet packages. We started getting this error after updating to Visual Studio 2019 version 16.5.0. Updating to the latest version of nuget.exe (5.4.0.6315) fixed the issue for us.
> 
> nuget.exe can be downloaded here: https://www.nuget.org/downloads.
ran nuget.exe - recommended latest v5.4.0 , still the same

**Comment by thpeder** [comment]
Hi @SirBETE, 
In order to unblock you, is it still crashing if you in your VSCode settings have?
```"al.useLegacyRuntime": false```

**Comment by SirBETE** [comment]
> Hi @SirBETE,
> In order to unblock you, is it still crashing if you in your VSCode settings have?
> `"al.useLegacyRuntime": false`

It was already set to false, so explicitly setting it in the settings made no difference.
But setting it to true, made the compiler **work**!

But now I got this strange warning:
{
	"resource": "./app.json",
	"owner": "_generated_diagnostic_collection_name_#1",
	"code": "AD0001",
	"severity": 4,
	"message": "Analyzer 'Microsoft.Dynamics.Nav.CodeCop.Design.Rule214BlindWriteLocalRecord' threw an exception of type 'System.InvalidCastException' with message 'System.InvalidCastException: Unable to cast object of type 'Microsoft.Dynamics.Nav.CodeAnalysis.BoundFieldAccess' to type 'Microsoft.Dynamics.Nav.CodeAnalysis.IInvocationExpression'.\r\n   at Microsoft.Dynamics.Nav.CodeCop.Design.Rule214BlindWriteLocalRecord.ProcessingInvocation(SyntaxNode syntaxNode, PooledDictionary`2 processInvocations, SyntaxNodeAnalysisContext syntaxNodeAnalysisContext)\r\n   at Microsoft.Dynamics.Nav.CodeCop.Design.Rule214BlindWriteLocalRecord.<>c.<Analyze>b__5_0(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Design\\Rule214BlindWriteLocalRecord.cs:line 66\r\n   at Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass52_1.<ExecuteSyntaxNodeAction>b__1() in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeAnalysis\\DiagnosticAnalyzer\\AnalyzerExecutor.cs:line 742\r\n   at Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze, Nullable`1 info) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeAnalysis\\DiagnosticAnalyzer\\AnalyzerExecutor.cs:line 1077'",
	"source": "AL",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1
}
x64 times ... :(


Any idea how I can solve those warnings?
Or what I need to do to not use the legacyruntime?

**Comment by thpeder** [comment]
I don't know how you can solve those warnings right away, but if you could boil it down to a small piece of self contained code that reproduces the stacktrace then we can investigate it.
For getting you away from using the legacy runtime, I have no clue at the moment.

Thanks for reporting the issue,
Thomas

**Comment by SirBETE** [comment]
> I don't know how you can solve those warnings right away, but if you could boil it down to a small piece of self contained code that reproduces the stacktrace then we can investigate it.
> For getting you away from using the legacy runtime, I have no clue at the moment.
> 
> Thanks for reporting the issue,
> Thomas

Hmmn, ok thanks!
I can at least use the new extension now. :)
I will try to create a project with those warnings.
TY for your assistance!

**Comment by atoader** [comment]
Were you able to reproduce this on a clean install? There are 2 runtimes that the language server uses:
- .NET Framework (useLegacy: true)
- .NET Core 3.0

The latter is a self-contained package and should work on any machine.

**Comment by SirBETE** [comment]
I could try, but I don't manage our environments. Doubt it will be allowed to let our internal IT create me a new server or spend much time on this to further test it ... :( sorry

I have these .NET core's installed:
![image](https://user-images.githubusercontent.com/34303738/81262409-1bfa9300-903e-11ea-987f-6385b0dfd9eb.png)
So would I need to download the 3.0 version?

**Comment by SirBETE** [comment]
Installed 3.0 (https://dotnet.microsoft.com/download/dotnet-core/3.0)
and removed the uselegacy setting, compiler DOESN'T crash! 😄 
awesome!!

This warning on app.json is still there though 😢 
{
	"resource": "/c:/project/x/app.json",
	"owner": "_generated_diagnostic_collection_name_#1",
	"code": "AD0001",
	"severity": 4,
	"message": "Analyzer 'Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions' threw an exception of type 'System.NullReferenceException' with message 'System.NullReferenceException: Object reference not set to an instance of an object.\r\n   at Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions.ExtractCountOfOptionsFromGlobalVar(ISymbol symbol) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Localizability\\Rule218TooltipMustBeFilledOnPageForFieldsAndActions.cs:line 286\r\n   at Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions.AnalyzeCaptions(SymbolAnalysisContext symbolAnalysisContext, SyntaxNode node, IdentifierNameSyntax identifierName, ISymbol symbol) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Localizability\\Rule218TooltipMustBeFilledOnPageForFieldsAndActions.cs:line 260\r\n   at Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions.AnalyzeModifiedControlsAndActions(SymbolAnalysisContext symbolAnalysisContext, SyntaxNode node, ISymbol symbol) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Localizability\\Rule218TooltipMustBeFilledOnPageForFieldsAndActions.cs:line 113\r\n   at Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions.AnalyzePageModifyFields(SymbolAnalysisContext symbolAnalysisContext, ImmutableArray`1 changeSymbols) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Localizability\\Rule218TooltipMustBeFilledOnPageForFieldsAndActions.cs:line 88\r\n   at Microsoft.Dynamics.Nav.CodeCop.Localizability.Rule218TooltipMustBeFilledOnPageForFieldsAndActions.AnalyzePage(SymbolAnalysisContext symbolAnalysisContext) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeCop\\Localizability\\Rule218TooltipMustBeFilledOnPageForFieldsAndActions.cs:line 56\r\n   at Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass49_2.<ExecuteSymbolActionsCore>b__0() in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeAnalysis\\DiagnosticAnalyzer\\AnalyzerExecutor.cs:line 655\r\n   at Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze, Nullable`1 info) in D:\\a\\1\\s\\source\\Prod\\Microsoft.Dynamics.Nav.CodeAnalysis\\DiagnosticAnalyzer\\AnalyzerExecutor.cs:line 1077'",
	"source": "AL",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1
}

A custom ruleset has been used, where rile 218 has been set to action : None.

**Comment by SirBETE** [comment]
Performance is now terrible :(
It takes very long (10s) to show the autocompletion, or to use the autoformat.
With the 4.0 compiler, it would be there instant, or with a small delay (1s).

**Comment by SirBETE** [comment]
@atoader 
Had to work today on a different project, where the compiler didn't crashed.
So I checked what's the difference between that and my current project (where the compiler crashes).
Saw "editor.codeLens": true, was different, changed it to false and restart VSC.
The compiler didn't crash!

What does that setting do?

**Comment by atoader** [comment]
Hi! There's a nice description of code lens here https://vscode.rocks/codelens/ . Having it enabled might be triggering some code paths that are causing the crash.

The other crash above seems to be coming from a CodeCop rule. Can you open another issue for that one? It seems that we are missing a null check .

**Comment by SirBETE** [comment]
Ok, I will check it and create a new issue. Thanks.

**Comment by SirBETE** [comment]
Looks like you fixed this problem. 
The compiler was updated, and it doesn't crash anymore, independent of whether codelens is active or not. 

 Awesome! :D
